modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/odata.es.js

```diff
@@ -0,0 +1,345 @@
+ * @param {object} oDataASTNode
+ * @returns String value of the internal name
+ */
+function getExpressionName(oDataASTNode) {
+	return oDataV4ParserNameMap[oDataASTNode.type];
+}
+
+/**
+ * Returns the next expression in the syntax tree that is not a grouping.
+ * @param {object} oDataASTNode
+ * @returns String value of the internal name of the next expression.
+ */
+const getNextNonGroupExpressionName = oDataASTNode => {
+	let returnValue;
+
+	if (oDataASTNode.value.type === 'BoolParentExpression') {
+		returnValue = getNextNonGroupExpressionName(oDataASTNode.value);
+	}
+	else {
+		returnValue = oDataASTNode.value.left ?
+			oDataASTNode.value.left :
+			oDataASTNode.value;
+	}
+
+	return getExpressionName(returnValue);
+};
+
+/**
+ * Checks if a grouping has different conjunctions (e.g. (x AND y OR z)).
+ * @param {boolean} lastNodeWasGroup
+ * @param {object} oDataASTNode
+ * @param {string} prevConjunction
+ * @returns boolean of whether a grouping has different conjunctions.
+ */
+function hasDifferentConjunctions({lastNodeWasGroup, oDataASTNode, prevConjunction}) {
+	return prevConjunction !== oDataASTNode.type && !lastNodeWasGroup;
+}
+
+/**
+ * Checks if the criteria is a group by checking if it has an `items` property.
+ * @param {object} criteria
+ */
+function isCriteriaGroup(criteria) {
+	return !!criteria.items;
+}
+
+/**
+ * Checks if the value is a certain type.
+ * @param {object} types A map of supported types.
+ * @param {*} value The value to validate.
+ * @returns {boolean}
+ */
+function isValueType(types, value) {
+	return Object.values(types).includes(value);
+}
+
+/**
+ * Checks if the group is needed; It is unnecessary when there are multiple
+ * groupings in a row, when the conjunction directly outside the group is the
+ * same as the one inside or there is no conjunction within a grouping.
+ * @param {boolean} lastNodeWasGroup
+ * @param {object} oDataASTNode
+ * @param {string} prevConjunction
+ * @returns a boolean of whether a group is necessary.
+ */
+function isRedundantGroup({lastNodeWasGroup, oDataASTNode, prevConjunction}) {
+	const nextNodeExpressionName = getNextNonGroupExpressionName(oDataASTNode);
+
+	return lastNodeWasGroup || oDataV4ParserNameMap[prevConjunction] === nextNodeExpressionName || !isValueType(CONJUNCTIONS, nextNodeExpressionName);
+}
+
+/**
+ * Removes a grouping node and returns the child node
+ * @param {object} oDataASTNode
+ * @param {string} prevConjunction
+ * @returns Object representing the operation inside the grouping
+ */
+function skipGroup({oDataASTNode, prevConjunction}) {
+	return {
+		lastNodeWasGroup: true,
+		oDataASTNode: oDataASTNode.value,
+		prevConjunction
+	};
+}
+
+/**
+ * Converts an OData filter query string to an object that can be used by the
+ * criteria builder
+ * @param {string} queryString
+ * @returns {object} Criteria representation of the query string
+ */
+function translateQueryToCriteria(queryString) {
+	let criteria;
+
+	try {
```

Comment:
Is it common that an error will be thrown? I normally don't see try/catch too often unless it is something async

Reply:
An error is common since the `odata-v4-parser` library throws an error if it encounters anything that can't be parsed.